### PR TITLE
fusion_info: show warning on status #403 ApiException, otherwise use …

### PR DESCRIPTION
**Pull request by andreepyro**
##### SUMMARY
This change introduces the following behaviour for module_fusion:
- if status #403 exception occurs, permission denied warning will be printed, but fusion_info module will return rest of the requested data
- if `purefusion.rest.ApiException` or `OperationException` occurs, it will be handled by error hook already used in toher modules
- code refactor: mainly changing `collection.items[ind] for ind in range(0, len(collection.items)):` pattern to `element for element in collection.items` (does the same but it is a lot easier to read)

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
plugins/fusion_info.py

### CODE REVIEW

**Comment by andreepyro** [comment]
Please don't merge until it's reviewed by the Pure Storage Fusion team

File: plugins/modules/fusion_info.py [hunk 1]
```diff
@@ -81,260 +81,459 @@
...
-                storage_service_name=storage_services.items[storserv].name
-            ).items
-        )
-    default_info["storage_classes"] = sclass
+    def inner(func):
+        def wrapper(module, fusion, *args, **kwargs):
+            try:
+                return func(module, fusion, *args, **kwargs)
+            except purefusion.rest.ApiException as exc:
+                if exc.status == 403:
```
Comment (genalt) [inline plugins/modules/fusion_info.py:105]: If it's not a permission denied, we should rethrow the exception, which will be handled a layer above

File: plugins/modules/fusion_info.py [hunk 2]
```diff
@@ -81,260 +81,459 @@
...
+        if exc.status == 403:
+            warning_api_exception("Users")
+
+    try:
+        protection_policies_num = len(
+            purefusion.ProtectionPoliciesApi(fusion).list_protection_policies().items
         )
-    default_info["role_assignments"] = assignments
+    except purefusion.rest.ApiException as exc:
+        if exc.status == 403:
```
Comment (genalt) [inline plugins/modules/fusion_info.py:167]: Please use a code from [http](https://docs.python.org/3/library/http.html#module-http)
In other places as well.

**Review by genalt — CHANGES_REQUESTED** [review]
Looks good. Just add several minor fixes. Thanks!

**Review by genalt — CHANGES_REQUESTED** [review]
There is a problem during testing:
```
fusion_info.py\", line 976, in generate_volumes_dict\nAttributeError: 'NoneType' object has no attribute 'name'\n"
```

**Comment by andreepyro** [comment]
@genalt Thanks for noticing!

File: plugins/modules/fusion_info.py [hunk 3]
```diff
@@ -350,469 +613,384 @@ def generate_array_dict(fusion):
...
-                            .mtu,
-                            "network_interface_group": endpoints.items[endpoint]
-                            .iscsi.discovery_interfaces[iface]
-                            .network_interface_groups[0]
-                            .name,
+                            "address": iface.address,
+                            "gateway": iface.gateway,
+                            "mtu": iface.mtu,
+                            "network_interface_groups": [
+                                nig.name for nig in iface.network_interface_groups
```
Comment (genalt) [inline plugins/modules/fusion_info.py:856]: there is a problem:
`fusion_info.py\", line 855, in generate_se_dict\nTypeError: 'NoneType' object is not iterable\n`

I would suggest just to add something like `nig.name for nig in (iface.network_interface_groups or [])`

**Review by genalt — CHANGES_REQUESTED** [review]
There is one bug, please fix it.

Also, we have a bunch of warnings:
```
[WARNING]: Cannot get Hardware Types in [default dict], reason: Permission denied
[WARNING]: Cannot get Appliances in [default dict], reason: Permission denied
[WARNING]: Cannot get Network Interface Groups in [default dict], reason: Permission denied
[WARNING]: Cannot get Network Interfaces in [default dict], reason: Permission denied
[WARNING]: Cannot get [hardware dict], reason: Permission denied
[WARNING]: Cannot get [interfaces dict], reason: Permission denied
[WARNING]: Cannot get [arrays dict], reason: Permission denied
[WARNING]: Cannot get [network_interface_groups dict], reason: Permission denied
```
I see there are duplicates, but just with a different format. It looks like the exception was handled in a particular method that generates a single default dict (using decorator), but then you handle it again

**Review by genalt — APPROVED** [review]

**Pull request merged**